### PR TITLE
Insertion sort update

### PR DESCRIPTION
**Pull request by Goober5000**
Turn `insertion_sort` into a templated function that uses assignment, so that it can be used with non-POD types.  Also make the sort do lazy-copying of elements and make a few other tweaks to the function.

### CODE REVIEW

**Comment by z64555** [comment]
Why is `windebug.cpp` being compiled by the linux builds, anyway?

**Comment by z64555** [comment]
Ok, as @BMagnu suggested in discord, we can try `#include <windows.h> //NOLINT`

If that doesn't work, we can also try guarding the .cpp with `#ifdef _WIN32`.

Probably the better solution would be to update the regex to ignore `globalincs/windebug.cpp` in `ci/linux/clang-tidy.sh`

**Comment by Goober5000** [comment]
Ok, regex updated and sort tweaked again.

File: code/globalincs/systemvars.h [hunk 1]
```diff
@@ -213,7 +213,56 @@ int current_detail_level();
 
 
 // Goober5000
-void insertion_sort(void *array, size_t array_size, size_t element_size, int (*fncompare)(const void *, const void *));
-
+// (Taylor says that for optimization purposes malloc/free should be used rather than vm_malloc/vm_free here)
+// NOTE: j *must* be a signed type because j reaches -1 and j+1 must be 0.  So we might as well make them ints
+// and make array_size an int as well.
+template <typename T>
+void insertion_sort(T* array_base, int array_size, int (*fncompare)(const T*, const T*))
```
Comment (z64555) [inline code/globalincs/systemvars.h:220]: Recommend documenting the parameters, especially the `fncompare`.  I'm assuming it's formed similarly to std::operators, i.e. `bool operator>(left, right)`

Bonus points for doxygen format.

**Review by z64555 — APPROVED** [review]

File: code/globalincs/systemvars.h (hunk 1 above)
Comment (Goober5000) [reply code/globalincs/systemvars.h:220]: k

**Pull request merged**